메인페이지 기능 구현

42KIM opened the pull request:
## 메인페이지 create-panel 기능 구현

### DB group_member 테이블 수정

+  create-panel에서 스터디 이름을 출력해주기 위해, study_group 테이블 접근 없이 group_member 테이블에서 모두 처리할 수 있도록 하기 위해 group_member 테이블에 ```study_group_name``` 컬럼을 생성하여 study_group 테이블의 ```name`` 컬럼을 참조하도록 DB 수정
+ 이에 따라 ```routes/create.js```에서 스터디 생성하기 쿼리문 수정



### 유저가 가입한 스터디가 랜덤으로 2개 리스트에 출력

+ 로그인된 유저의 경우만 해당 부분이 렌더링 됨
+ 이 부분 구현을 위해 routes/index.js의 ```router.get('/')```을 수정함  
  로그인 되지 않은 유저가 메인페이지에 접근했을 때, 불필요한 DB 접근을 방지하기 위해, 로그인 된 경우/로그인 되지 않은 경우를 분기로 나눴음. 로그인 된 유저의 경우, ```group_member``` 테이블에 접근하여 랜덤하게 2개 추출하고 해당 내용을 ```index.ejs```의 ```myGroups``` 변수에 담아서 전달.
+ 비로그인 유저의 경우 위의 과정은 생략하고 ```study_group```에서 search-list에 렌더링될 데이터만 가져옴. (```mainGroups``` 변수에 담아서 ```index.ejs```에 전달)



### create-panel의 내가 가입한 스터디 클릭 시, 해당 스터디 소개페이지로 이동

+ ```public/js/create-panel.js``` 파일 생성
+ 클릭 시, hidden input에 담겨있던 해당 스터디의 study_group_id를 ``` `http://localhost:3000/group/info/${groupID}`;```로 보내주는 ```toGroupInfo()``` 함수 생성

youngToMaturity committed:
feat: mypage 로드시, 로그인 된 user가 가입/생성한 그룹의 데이터를 fetch

각 스터디 클릭시 group-edit에 group_id를 넘길 수 있도록 임시로 실행하고자 하는 과정. (mypage가 완성되면 수정 예정)

youngToMaturity committed:
feat: fetch를 통해 받은 study_gorup_id를 각 group-list__edit-btn의 value로 지정

[임시]
! issue : 각 lead-group-list의 group-list__item의 갯수를 어떻게 지정할 것인가? - user마다 운영하고 있는 스터디 그룹의 갯수가 다를것이기 때문에
! issue: group-list__item에 해당 study_group_id까지는 지정할 수 있겠지만, 해당 div의 내용들을 어떻게 바꿀 것인가? (fetch속의 fetch?)

youngToMaturity committed:
feat : 관리 버튼 클릭시 각 item에 해당하는 study-group-id post

관리 버튼 클릭시 group-list-hidden__input에 저장된 study-group-id가 post된 후 group-edit이 render되도록 함

youngToMaturity committed:
feat: '관리' 버튼 클릭시 post로 group-edit.ejs에 선택된 그룹의 정보 넘기기

db.query를 활용하여 submit된 id와 같은 study_group을 찾고, 해당 그룹에 속해있는 group_member도 result로 저장. 하지만 ejs에 대한 이해도 부족으로 이 값들을 ejs 상에 표시하는것 미완성

42KIM committed:
feat: Load lead-group list by ssr+csr

youngToMaturity committed:
style :  ejs 적용과정 공유를 위한 커밋

youngToMaturity committed:
feat : group-edit 페이지에서 기존의 스터디 그룹 정보를 기본값으로 표시해주기

수정하기 페이지에 들어가는 경우, 해당 스터디의 기존 정보를 page에 보이도록 하였습니다. ( 중분류, 멤버 미구현 )

youngToMaturity committed:
feat: group-edit 페이지 상의 중분류 select를 수정하고자 하는 스터디 그룹의 기존 값으로 지정

route/mypage.js의 /fetch를 /fetch-group으로 수정
js/group-edit.js의 categoryChange(e)에 있던 중분류 list를 전역으로 선언하여 DOMContentLoad시와 change시 모두 활용할 수 있도록 함.
categotyChange(e) 함수내부의 중분류 Element들의 생성과 지정을 setSub()함수로 처리하였고, hidden Input을 통해 확인할 수 있는 render시에 받은 sub과 같은 경우 selected 옵션을 추가.

swprk committed:
feat: Render 'group-info' with data

42KIM committed:
refactor: Change MySQL BETWEEN to LIMIT

42KIM committed:
style: Change variable name

youngToMaturity committed:
feat: 멤버의 수만큼 div를 생성하여 닉네임과 강퇴버튼을 부여

강퇴 버튼의 value에 group-member의 id를 지정했고, 닉네임은 각 회원의 닉네임으로 지정했다. 현재는 임시로 방장도 출력했지만 추후에는 방장을 제외하고 출력하도록 처리할 것

42KIM committed:
feat: 마이페이지 다중쿼리, 서브쿼리 추가

swprk committed:
fix: Handle exception for req.user

42KIM committed:
feat: 무한스크롤 기능 구현, 더보기 버튼 삭제

42KIM committed:
feat: Enable create-panel render, link

youngToMaturity committed:
feat : 수정완료 버튼 클릭시 db 수정, 예외처리 두가지 적용

만약 db상에 수정하고자 하는 이름이 있고 && 그 이름이 현재 수정하고자 하는 스터디 그룹의 이름이 아니라면 수정불가.
만약 현재 스터디 인원수보다 수정하고자 하는 스터디 정원이 더 적다면 수정불가.
그 외의 상황은 수정 가능

youngToMaturity committed:
feat: 강퇴 버튼 클릭시 해당 멤버 강퇴 기능 구현

group-edit.ejs - form 위치 수정 (css 재지정 필요)
routes/mypage.js - kickoutForm post 처리, 방장인 경우(추후 방장은 members에 보이지 않도록 처리 요망) 강퇴안됨
js/mypage.js - Kickout시 방생되는 status 2개 추가, 기존의 complete은 edit으로 수정 ?: 각 case 처리 이후에 redirect할 수 있을까?
js/group-edit.js 강퇴 버튼 클릭시 kickout 폼 submit

42KIM committed:
feat: 메인페이지 search-list 구현

42KIM committed:
feat: DB에 생성된 스터디가 없을 때 에러 처리

42KIM committed:
fix: 데이터 없는 경우 분기로 생성

youngToMaturity committed:
Merge remote-tracking branch 'origin/ssr+csr' into group-edit

# Conflicts:
#	public/js/mypage.js
#	routes/mypage.js
#	views/includes/mypage/lead-group-list.ejs

swprk committed:
feat: Add comment section to group-info

swprk committed:
fix: Modify sql query

swprk committed:
fix: Resolve conflict

42KIM committed:
style: Change button tag to input tag, remove classNames

42KIM committed:
feat: Enable mainpage category changing and search-list rendering

42KIM commented:
## 메인페이지 Category-filter 및 Search-list 구현

### 진행 흐름

+ 최초 메인페이지 접속 시, DB에 존재하는 스터디 중 create_date를 기준으로 10개가 SSR됩니다.
+ DB에 스터디가 존재하지 않는 경우 ''현재 모집 중인 스터디가 없습니다..."라는 메시지가 search-list에 출력됩니다. (해당 카테고리에 스터디가 없는 경우도 마찬가지입니다.)
+ 카테고리를 변경하는 경우의 흐름은 다음과 같습니다.
  + 버튼을 클릭하면, 해당 버튼의 value (카테고리 명)가 ```localhost:3000/?main=[이 자리]```에 담겨 페이지가 redirect 됩니다. 
  + 클릭한 카테고리에 해당되는 스터디 10개를 리스트에 다시 SSR 해주는 방식을 사용하기 때문입니다.
  + ```router/index.js```에서는 쿼리스트링을 해석하여, 해당 카테고리의 스터디를 DB로부터 가져와 ejs 템플릿에 전달합니다.
  + 메인페이지 화면이 새로고침 되면서, main-category-filter에 '선택된 버튼'을 표시하고, 해당 중분류 버튼을 출력하기 위한 함수가 작동합니다.   ```public/js/main-category-filter.js```의 ```loadCategory()```함수. 추후 설명.
  + 중분류를 선택하는 경우 ```localhost:3000/?main=대분류&sub=[이 자리]```에 버튼의 value 값이 담겨 전달되고 마찬가지 과정이 반복됩니다.

+ 🚩 현재 리스트 추가 로드 기능(무한 스크롤)은 구현되지 않았습니다. 메인페이지 구현 방법이 확정되면 마이페이지에서 사용한 로드 기능을 똑같이 사용할 예정입니다.



### 변경된 사항

자세한 로직 설명에 앞서, 기존 코드와 크게 변경된 부분을 짚고 넘어가야 할 것 같습니다.

#### 1. main-category-filter 분리

기존에는 ```category-filter.js```와 ```category-filter.ejs```가 메인페이지와 스터디 생성 페이지에서 함께 사용되고 있었습니다. 그러나 메인페이지에서 카테고리 변경에 따른 데이터 요청 및 렌더링 기능을 구현하기에 기존 방법은 적합하지 않아서, ```main-category-filter.js```와 ```main-category-filter.ejs``` 파일을 새롭게 생성했습니다.  
🚩 이 방법이 확정되면, 기존 ```category-filter```의 디렉토리를 index에서 create쪽으로 변경해줘야 합니다. 

#### 2. 카테고리 버튼을 ```button```태그에서 ```input```태그로 수정했습니다.

```main-category-filter.ejs```에서 카테고리 버튼의 태그를 변경했습니다.  

기존:

```html
<button class="cat-filter-main__btn officiary">공무원</button>

<button class="cat-filter-sub__btn officiary">9급</button>
```

변경:

```html
<input class="cat-filter-main__btn" type="button" value="공무원" />

<input class="cat-filter-sub__btn 공무원" type="button" value="9급" />
```

type을 button으로 주면 CSS도 동일하게 적용됩니다.

기존 방식에서는 class명과 실제 표기되는 한글 text를 일일이 비교해야하는 불편함이 있었고, 다른 페이지나 엘레먼트로 값을 전달할 때도 ```.textContent```등을 사용하여 값을 얻기 위한 과정이 번거로웠습니다. 

가독성 측면에서도 개선되었다고 느껴지고, 불필요한 영어 class 명을 쓰지 않아도 되는 것 같습니다.

🚩 스터디 생성 페이지도 이 방법을 적용한다면, 값을 전달할 때의 코드를 일부 변경하면 됩니다.

#### 3. routes/mypage, routes/index의 isLoggedIn 함수 제거

기존에 해당 페이지 접근 시 사용자의 로그인 여부를 체크하던 ```isLoggedIn```함수의 사용을 중지했습니다. 기존에는 라우터에서 분기가 없었기 때문에 로그인 여부에 따라 true/false를 return하여 ejs에 전달해주던 함수가 유용했지만,

 지금은 라우터에서 다양한 데이터를 함께 담아 보내기 때문에 굳이 함수를 사용해서 true, false 값을 **재차 구해서 전달**할 필요가 없었습니다. 이미 로그인 된 경우, 로그인 되지 않은 경우를 분기로 나눴기 때문입니다. 따라서 각 분기에 맞게 true, false를 직접 전달하는 게 불필요한 함수 호출을 줄일 수 있다고 생각했습니다.

#### 4. routes/index의 라우터 수정

이제 메인페이지는 로그인/비로그인에 따라 header와 create-panel이 바뀌고,

유저가 가입한 스터디에 따라 create-panel이 바뀌고,

생성되는 스터디에 따라 search-list가 바뀌고,

선택하는 카테고리에 따라 category-filter가 바뀌어

매번 다른 ejs 템플릿을 SSR하여 메인페이지에 응답해야 합니다. 따라서 ```router.get('/')```가 바뀌었습니다. 대략적인 흐름은 다음과 같습니다.

+ 현재 URL의 쿼리스트링으로 선택된 카테고리를 판별하여 그에 맞는 쿼리를 DB에 요청합니다. 가져온 데이터를 ejs에 전달합니다.  (파라미터가 둘 다 없는 경우 최초 접속(= '전체' 카테고리) 입니다.)
+ search-list 출력에 관해서는 '최초 접속', '대분류만 선택한 경우', '중분류도 선택한 경우' 총 세 가지 경우의 수가 있습니다.
+ create-panel에 회원이 가입한 스터디를 랜덤하게 2개 선별하여 전달하는 쿼리도 추가되었습니다.



#### 5. main-category-filter.js 카테고리 선택 / 변경 / 렌더링 구현

기존에 handleMain, handleSub 함수로 클릭된 element의 class만 바꿔주던 category-filter로는 클릭마다 페이지가 새로고침으로 초기화되는 SSR 방식을 적용하기가 어려웠습니다. 

바뀐 흐름은 다음과 같습니다.

+ 새로운 카테고리 버튼을 클릭하면 : ```changeCategory() 함수```
  + 해당 버튼이 main-category 버튼일 경우, 기존 ```active-main``` 클래스가 있던 버튼의 클래스를 제거하고 현재 클릭된 버튼의 value를 쿼리스트링에 담아 페이지를 새로고침 합니다.
  + 해당 버튼이 sub-category 버튼일 경우, 현재 ```active-main``` 클래스가 있는 '대분류' 버튼을 찾고, 선택된 중분류 버튼의 value를 함께 쿼리스트링에 담아 새로고침 합니다.
+ 페이지가 새로고침 되면 : ```loadCategory() 함수```
  + 먼저 ```URLSearchParams```로 쿼리스트링의 파라미터 값들을 확인합니다.
  + 메인페이지 첫 접속이거나, '전체' 카테고리를 선택한 경우 ```복잡한 선택자 문법```을 사용하여 '전체' 대분류 버튼을 찾아 ```active-main``` 클래스를 붙여줍니다.
  + 그 외의 카테고리를 선택한 경우 대분류를 바꾼 경우, 중분류를 바꾼 경우 두 가지 경우의 수가 있습니다.
  + 대분류를 변경한 경우 마찬가지로 복잡한 선택자 문법을 사용하여 현재 쿼리스트링으로 전달된 값에 해당되는 버튼을 찾아 ```active-main```을 붙여주고, 그 대분류에 해당되는 중분류 버튼들에 ```show-sub```클래스를 붙여서 화면에 그려줍니다.
  + 중분류를 변경한 경우 : ```if (curSubCat)```. 마찬가지로 쿼리스트링으로 전달된 중분류 값에 해당되는 버튼을 찾아서 ```active-sub``` 클래스를 붙여줍니다.
  + 이러한 과정을 통해서 버튼 클릭마다 페이지가 새로고침 되어도 '클릭한 버튼'이 초기화되지 않고 화면에 제대로 표시될 수 있었습니다.



제가 떠올린 카테고리 버튼 클릭에 따라서 새로운 search-list를 출력해주고, 페이지가 새로고침 되어도 선택한 버튼이 눌려져있는 상태를 화면에 그려주는 방법은 이렇습니다. (기본적으로는 SSR, 데이터가 추가 로드될 때는 CSR)

좀 더 세부적으로 기술적인 효율성을 높이거나, 비효율적인 코드를 줄일 수 있는 방법을 함께 고민해보면 좋을 것 같습니다.

ghost commented:
> 기존에 handleMain, handleSub 함수로 클릭된 element의 class만 바꿔주던 category-filter로는 클릭마다 페이지가 새로고침으로 초기화되는 SSR 방식을 적용하기가 어려웠습니다.

기존에 영수님이 작성하신 `category-filter.js`는 새로고침 없이 버튼에 `active`를 주는 코드였는데, 지금은 카테고리 버튼마다 리다이렉션이 되어야 하는 상황이니까, DOM 조작 없이 이미 `active`가 추가된 EJS를 SSR 하는 방법도 있을 것 같습니다. 근데 이렇게 하면 분기가 복잡해지니까 어렵다고 하신 걸로 이해되네요.

한땀한땀 고민하신 흔적이 보입니다! 꼼꼼히 읽어봤는데 지금은 비효율적이라고 생각되는 코드는 없습니다! 저희 수준에서 이 정도면 최대치인 것 같고 앞으로 더 좋은 방법을 배워야 그 때 개선할 수 있을 것 같습니다.

sanoopark commented on views/includes/index/main-category-filter.ejs:
```diff
@@ -0,0 +1,37 @@
+<% if (path === '/create') { %>
+<div class="group-create-category__main">대분류</div>
+<% } %>
```
`index/category-filter.ejs`와 `create/category-filter.ejs`로 나누면 이 부분은 삭제해야 되는 건가요?

42KIM replied:
아 네 그렇겠네요! 
create 라우터에서는 어느 부분을 수정해야할지 꼼꼼하게 살펴보지 않았습니다. 나누게 된다면 어디어디를 바꿔야할지 잘 찾아봐야할 것 같네요.

42KIM committed:
Merge branch 'group-info' into mainpage

42KIM committed:
refactor: Change button tags to input in category-filter

42KIM committed:
fix: 관리 버튼 클릭 시 스터디 관리 페이지로 이동 못하는 문제 해결

42KIM committed:
Merge branch 'group-info' into mainpage

42KIM committed:
feat: Add search-list link to group-info page

youngToMaturity committed:
feat : 수정이 정상적으로 처리된 경우에는 mypage로, 그렇지 않은 경우 group-edit 페이지로 redirect

lead-group-list에서 관리 버튼을 클릭하는 경우 form을 submit하여 router.post하던 방식에서 location.href를 통해 group-edit/:id로 이동시킨 후 router.get()을 사용하여 render

youngToMaturity committed:
feat : group-edit의 과정 수정 - 스터디명 중복체크 이후에 수정완료 버튼 활성화, 수정완료 버튼 활성화 이후 스터디명 수정시 다시 비활성화

youngToMaturity committed:
fix : 수정하기 버튼명을 중복체크로 수정
